datumaro/plugins/data_formats/arrow/arrow_dataset.py

```diff
@@ -0,0 +1,203 @@
+from typing import Dict, List, Optional, Tuple, TypeVar, Union, overload
+
+import numpy as np
+import pyarrow as pa
+
+PathLike = TypeVar("PathLike", str, os.PathLike, None)
+
+
+def _binary_search(arr, x, low=None, high=None, fn=None):
+    low = 0 if low is None else low
+    high = len(arr) - 1 if high is None else high
+    if fn is None:
+
+        def fn(arr, x):
+            return arr[x]
+
+    while low <= high:
+        mid = (high + low) // 2
+        val = fn(arr, mid)
+        if val > x:
+            high = mid - 1
+        else:
+            low = mid + 1
+    return high
+
+
+class ArrowDatasetIterator(Iterator):
+    def __init__(self, dataset: "ArrowDataset"):
+        self.dataset = dataset
+        self.index = 0
+
+    def __next__(self) -> Dict:
+        if self.index >= len(self.dataset):
+            raise StopIteration
+        item = self.dataset[self.index]
+        self.index += 1
+        return item
+
+
+class ArrowDataset:
+    def __init__(
+        self,
+        files: Union[PathLike, List[PathLike]],
+        keep_in_memory: bool = False,
+    ):
+        assert files
+        if not isinstance(files, (list, tuple)):
+            files = [files]
+
+        self._files = files
+        self._keep_in_memory = keep_in_memory
+
+        self.unload_arrows()
+        self.load_arrows()
+
+    @property
+    def table(self) -> pa.Table:
+        self.load_arrows()
+        return self.__table
+
+    @property
+    def column_names(self) -> List[str]:
+        return self.table.column_names
+
+    @property
+    def num_columns(self) -> int:
+        return self.table.num_columns
+
+    @property
+    def num_rows(self) -> int:
+        return self.table.num_rows
+
+    @property
+    def shape(self) -> Tuple[int]:
+        return self.table.shape
+
+    @property
+    def metadata(self) -> Optional[Dict[bytes, bytes]]:
+        return self.table.schema.metadata
+
+    def __deepcopy__(self, memo) -> "ArrowDataset":
+        args = [self._files, self._keep_in_memory]
+        copied_args = copy.deepcopy(args, memo)
+        return self.__class__(*copied_args)
+
+    def __len__(self) -> int:
+        return self.table.num_rows
+
+    def __iter__(self):
+        return ArrowDatasetIterator(self)
+
+    def unload_arrows(self) -> None:
+        self.__table: pa.Table = None
+        self.__batches: List[pa.RecordBatch] = []
+        self.__offsets: List[int] = []
+
+    def load_arrows(self) -> None:
+        if self.__table is not None:
+            return
+
+        table = None
+        for file in self._files:
+            if not os.path.exists(file):
+                raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), file)
+
+            if self._keep_in_memory:
+                stream = pa.input_stream(file)
+            else:
+                stream = pa.memory_map(file)
+            opened_stream = pa.ipc.open_stream(stream)
+            table_ = opened_stream.read_all()
```

Comment:
What is the difference between `self._keep_in_memory=True or False`? I think that there seems no difference after `table_ = opened_stream.read_all()` because it seems to load everything on the memory from the file.